### ISSUE TITLE
Compile Failure "...unexec: unrecognized section..."   on MacOS El Capitán

### ISSUE THREAD

**Issue by pazthor** [Wilfred/remacs#17]
Its my first try to compile Remacs,  I follow the intruction but this problem I don't understand, I think is a version of packages in OS X. 

```
--- List of All Regions ---
   address     size prot maxp
--- List of Regions to be Dumped ---
   address     size prot maxp
--- Header Information ---
Magic = 0xfeedfacf
CPUType = 16777223
CPUSubType = -2147483645
FileType = 0x2
NCmds = 28
SizeOfCmds = 4152
Flags = 0x00800085
Highest address of load commands in input file: 0x10085e000
Lowest offset of all sections in __TEXT segment:   0x2700
--- List of Load Commands in Input File ---
# cmd              cmdsize name                address     size
0 LC_SEGMENT_64          72 __PAGEZERO                0 0x100000000
1 LC_SEGMENT_64         952 __TEXT           0x100000000 0x29f000
                           __text           0x100002700 0x231c08
                           __stubs          0x100234308    0xc12
                           __stub_helper    0x100234f1c   0x142e
                           __const          0x100236350  0x2f930
                           __cstring        0x100265c80  0x1552b
                           __gcc_except_tab 0x10027b1ac   0x3778
                           __objc_methname  0x10027e924   0x3612
                           __objc_classname 0x100281f36    0x114
                           __objc_methtype  0x10028204a   0x1653
                           __unwind_info    0x1002836a0   0x3910
                           __eh_frame       0x100286fb0  0x18048
2 LC_SEGMENT_64        1672 __DATA           0x10029f000 0x403000
                           __nl_symbol_ptr  0x10029f000     0x10
                           __got            0x10029f010    0x278
                           __la_symbol_ptr  0x10029f288   0x1018
                           __const          0x1002a02a0   0xfcc0
                           __cfstring       0x1002aff60    0x880
                           __objc_classlist 0x1002b07e0     0x78
                           __objc_catlist   0x1002b0858      0x8
                           __objc_protolist 0x1002b0860     0x28
                           __objc_imageinfo 0x1002b0888      0x8
                           __objc_const     0x1002b0890   0x3bd0
                           __objc_selrefs   0x1002b4460   0x10c0
                           __objc_classrefs 0x1002b5520    0x218
                           __objc_superrefs 0x1002b5738     0x60
                           __objc_ivar      0x1002b5798    0x200
                           __objc_data      0x1002b5998    0x4b0
                           __data           0x1002b5e50 0x340fc8
                           __thread_vars    0x1005f6e18     0x90
                           __thread_data    0x1005f6ea8     0xc8
                           __bss            0x1005f6f70  0x89ac8
                           __common         0x100680a40  0x20791
3 LC_SEGMENT_64          72 __LINKEDIT       0x1006a2000 0x1bc000
4 LC_DYLD_INFO_ONLY      48
5 LC_SYMTAB              24
6 LC_DYSYMTAB            80
7 LC_LOAD_DYLINKER       32
8 LC_UUID                24
9 LC_VERSION_MIN_MACOSX      16
10 LC_SOURCE_VERSION      16
11 LC_MAIN                24
12 LC_LOAD_DYLIB          88
13 LC_LOAD_DYLIB          88
14 LC_LOAD_DYLIB          64
15 LC_LOAD_DYLIB          88
16 LC_LOAD_DYLIB          88
17 LC_LOAD_DYLIB          56
18 LC_LOAD_DYLIB          56
19 LC_LOAD_DYLIB          56
20 LC_LOAD_DYLIB          48
21 LC_LOAD_DYLIB         104
22 LC_LOAD_DYLIB         104
23 LC_LOAD_DYLIB          96
24 LC_LOAD_DYLIB          96
25 LC_LOAD_DYLIB          56
26 LC_FUNCTION_STARTS      16
27 LC_DATA_IN_CODE        16
0x101efc080 (sz:   0x3f24/  0x3f28)
0x101e00000 (sz:  0x1b4c5/ 0xfc080)
0x1018fc080 (sz:   0x3f24/  0x3f28)
0x101800000 (sz:  0xdcdca/ 0xfc080)
0x101dfc080 (sz:   0x3f24/  0x3f28)
0x101d00000 (sz:  0x439c5/ 0xfc080)
0x101cfc080 (sz:   0x3f24/  0x3f28)
0x101c00000 (sz:  0xe2d60/ 0xfc080)
0x1027f8000 (sz:   0x75e0/  0x7fa0)
0x102000000 (sz: 0x75bdff/0x7f8000)
0x107ff8000 (sz:   0x16b6/  0x7fa0)
0x107800000 (sz: 0x1693fd/0x7f8000)
0x1077f8000 (sz:   0x1c86/  0x7fa0)
0x107000000 (sz: 0x1c63ff/0x7f8000)
0x106ff8000 (sz:   0x5eea/  0x7fa0)
0x106800000 (sz: 0x5ec7ff/0x7f8000)
0x100ca9000 (sz:        0/  0x1000)
--- Load Commands written to Output File ---
Writing segment __PAGEZERO       @        0 (       0/0x100000000 @          0)
Writing segment __TEXT           @        0 (0x29f000/0x29f000 @ 0x100000000)
Writing segment __DATA           @ 0x29f000 (0x403000/0x403000 @ 0x10029f000)
        section __nl_symbol_ptr  at 0x29f000 - 0x29f010 (sz:     0x10)
        section __got            at 0x29f010 - 0x29f288 (sz:    0x278)
        section __la_symbol_ptr  at 0x29f288 - 0x2a02a0 (sz:   0x1018)
        section __const          at 0x2a02a0 - 0x2aff60 (sz:   0xfcc0)
        section __cfstring       at 0x2aff60 - 0x2b07e0 (sz:    0x880)
        section __objc_classlist at 0x2b07e0 - 0x2b0858 (sz:     0x78)
        section __objc_catlist   at 0x2b0858 - 0x2b0860 (sz:      0x8)
        section __objc_protolist at 0x2b0860 - 0x2b0888 (sz:     0x28)
        section __objc_imageinfo at 0x2b0888 - 0x2b0890 (sz:      0x8)
        section __objc_const     at 0x2b0890 - 0x2b4460 (sz:   0x3bd0)
        section __objc_selrefs   at 0x2b4460 - 0x2b5520 (sz:   0x10c0)
        section __objc_classrefs at 0x2b5520 - 0x2b5738 (sz:    0x218)
        section __objc_superrefs at 0x2b5738 - 0x2b5798 (sz:     0x60)
        section __objc_ivar      at 0x2b5798 - 0x2b5998 (sz:    0x200)
        section __objc_data      at 0x2b5998 - 0x2b5e48 (sz:    0x4b0)
        section __data           at 0x2b5e50 - 0x5f6e18 (sz: 0x340fc8)
unexec: unrecognized section __thread_vars in __DATA segment
make[1]: *** [bootstrap-emacs] Error 1
make: *** [src] Error 2


```

**Comment by rogermarlow** [comment]
Exact same problem here on MacOS 10.12.2
Last step in the build appears to be
```
./temacs --batch --load loadup bootstrap
[ lines deleted ]
Loading /Users/rogermarlow/github/remacs/lisp/cus-start.el (source)...
Loading /Users/rogermarlow/github/remacs/lisp/tooltip.el (source)...
Finding pointers to doc strings...
Finding pointers to doc strings...done
Dumping under the name emacs
--- List of All Regions ---
   address     size prot maxp
[ see preview commend for rest of dump ]
```

**Comment by rogermarlow** [comment]
As the error message says, it doesn't recognize the __thread_vars section. By the way __thread_data is also listed in the commands it is expecting to see and that isn't in the code either. So wild guess at the following hack:
`src/unexmacosx.c`
```
--- a/src/unexmacosx.c
+++ b/src/unexmacosx.c
@@ -929,7 +929,9 @@ copy_data_segment (struct load_command *lc)
           || strncmp (sectp->sectname, "__mod_init_func", 16) == 0
           || strncmp (sectp->sectname, "__mod_term_func", 16) == 0
           || strncmp (sectp->sectname, "__static_data", 16) == 0
-          || strncmp (sectp->sectname, "__objc_", 7) == 0)
+          || strncmp (sectp->sectname, "__objc_", 7) == 0
+          || strncmp (sectp->sectname, "__thread_data", 13) == 0
+          || strncmp (sectp->sectname, "__thread_vars", 13) == 0)
    {
      if (!unexec_copy (sectp->offset, old_file_offset, sectp->size))
        unexec_error ("cannot copy section %.16s", sectp->sectname);
```
at least gets you past the the dumping out temacs stage. (Ensure you do `make clean` first). Then we are on to the next problem. Built in `consp` not defined. Don't think that is a result of my hack because I don't see a file of that name or any function definition for it.

**Comment by Wilfred** [comment]
Great debugging @rogermarlow! :)

Note that `consp` is defined in cons.rs. Are you sure you have a fresh compile of the Rust code?

**Comment by kennethkalmer** [comment]
For those following along, I applied the diff above, did a `make clean` and ended up with this error during `make`

```
4480 unused bytes follow Mach-O header
89147 pure bytes used
mv -f remacs bootstrap-emacs
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../lisp compile-first EMACS="../src/bootstrap-emacs"
  ELC      emacs-lisp/macroexp.elc
Invalid function: eq
make[2]: *** [emacs-lisp/macroexp.elc] Error 255
make[1]: *** [bootstrap-emacs] Error 2
make: *** [src] Error 2
```

**Comment by rogermarlow** [comment]
Yep, I'm getting exactly the same. I've issued PR #37 for the LC segment fix which I think is a separate issue to what we are seeing now. Let's open new issue if we can't get `eq` to work.

**Comment by rogermarlow** [comment]
Moved to issue #38

**Comment by Wilfred** [comment]
I think the LC segment issue was fixed in #37, and the issue with `eq` is being discussed in #38. I don't think there's anything left to here. Let me know if you disagree :).

**Comment by pazthor** [comment]
I'm agreed! :D
